README.md
# Flutter Walkthrough Screen

A flutter package which help developer in creating a customize onboarding screens of their app.

 ![Bitbucket open issues](https://img.shields.io/bitbucket/issues-raw/champ96k/flutter_walkthrough_screen)


## Screenshots

![](https://i.ibb.co/L05Kg47/Walkthrough.png)


## Usage


[Example](https://github.com/champ96k/flutter_walkthrough_screen/tree/main/example)

To use this package :

* add the dependency to your [pubspec.yaml](https://github.com/champ96k/flutter-onboarding-screen/blob/master/pubspec.yaml) file.

```yaml
  dependencies:
    flutter:
      sdk: flutter
    flutter_walkthrough_screen:
```

### How to use

```dart
class TestScreen extends StatelessWidget {
    /*here we have a list of OnbordingScreen which we want to have, each OnbordingScreen have a imagePath,title and an desc.
      */
  final List<OnbordingData> list = [
      OnbordingData(
        image: AssetImage("images/pic1.png"),
        titleText:Text("This is Title1"),
        descText: Text("This is desc1"),
      ),
       OnbordingData(
        image: AssetImage("images/pic2.png"),
        titleText:Text("This is Title2"),
        descText: Text("This is desc2"),
      ),
       OnbordingData(
        image: AssetImage("images/pic3.png"),
        titleText:Text("This is Title3"),
        descText: Text("This is desc4"),
      ),
       OnbordingData(
        image: AssetImage("images/pic4.png"),
        titleText:Text("This is Title4"),
        descText: Text("This is desc4"),
      ),
    ];

  @override
  Widget build(BuildContext context) {
    /* remove the back button in the AppBar is to set automaticallyImplyLeading to false
  here we need to pass the list and the route for the next page to be opened after this. */
    return IntroScreen(
                 onbordingDataList: list,
                 colors: [
                   //list of colors for per pages
                   Colors.white,
                   Colors.red,
                 ],
                 pageRoute: MaterialPageRoute(
                   builder: (context) => NextPage(),
                 ),
                 nextButton: Text(
                   "NEXT",
                   style: TextStyle(
                     color: Colors.purple,
                   ),
                 ),
                 lastButton: Text(
                   "GOT IT",
                   style: TextStyle(
                     color: Colors.purple,
                   ),
                 ),
                 skipButton: Text(
                   "SKIP",
                   style: TextStyle(
                     color: Colors.purple,
                   ),
                 ),
                 selectedDotColor: Colors.orange,
                 unSelectdDotColor: Colors.grey,
               );
  }
}
```


# Customize your onboarding screen design


Background color             |  Background gradient color
:-------------------------:|:-------------------------:
![](https://i.ibb.co/87Yd36V/background-color.png)  |  ![](https://i.ibb.co/F8GYJSR/background-gradient-color.png)



Last page button             |  Next page button
:-------------------------:|:-------------------------:
![](https://i.ibb.co/5vLg0pG/last-page-button.png)  |  ![](https://i.ibb.co/qmVRSvV/next-page-button.png)


Skip button             |  Define next page route
:-------------------------:|:-------------------------:
![](https://i.ibb.co/f0qRn4C/skip-button.png)  |  ![](https://i.ibb.co/MM9f6Xx/page-route.png)


Selected dot Color             |  Unselected dot Color
:-------------------------:|:-------------------------:
![](https://i.ibb.co/6w344S3/selected-dot-Color.png)  |  ![](https://i.ibb.co/dtpjTD1/unselect-dot-Color.png)



### Created & Maintained By

[Tushar Nikam](http://tusharnikam.ml/)



<p><a href="https://www.twitter.com/champ_96k"><img src="https://img.shields.io/badge/twitter-%231DA1F2.svg?&style=for-the-badge&logo=twitter&logoColor=white" height=25></a> <a href="https://www.linkedin.com/in/tushar-nikam-a29a97131/"><img src="https://img.shields.io/badge/linkedin-%230077B5.svg?&style=for-the-badge&logo=linkedin&logoColor=white" height=25></a> <a href="https://medium.com/@champ96k"><img src="https://img.shields.io/badge/medium-%2312100E.svg?&style=for-the-badge&logo=medium&logoColor=white" height=25></a> <a href="https://champ96k.github.io"><img src="https://img.shields.io/badge/tusharnikam.ml-portfolio-orange" height=25></a> <a href="https://stackoverflow.com/users/11157840/tushar-nikam" target="_blank"><img src="https://logos-download.com/wp-content/uploads/2019/01/Stack_Overflow_Logo-700x283.png" alt="Stackoverflow" style="height: 41px !important;width: 174px !important;box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;-webkit-box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;"height=25 ></a>  <a href="https://www.buymeacoffee.com/champ96k" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png" alt="Buy Me A Coffee" style="height: 41px !important;width: 174px !important;box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;-webkit-box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;"height=25 ></a></p>


CHANGELOG.md
## [0.0.1] - Release

* A customizable flutter walkthrough screen library written in dart.

## [0.0.2] - Release

* A customizable flutter walkthrough screen library written in dart .

## [2.0.1] - Release

* A customizable flutter walkthrough screen library written in dart with a flutter null safe support.

## [2.0.2] - Release
    - Improve documentation and format code

lib/walkthrough_screen.dart
library walkthrough_screen;
import 'package:flutter/material.dart';

class OnbordingData extends StatelessWidget {
  final ImageProvider<Object> image;
  final Widget titleText;
  final Widget descText;
  final double imageHeight;
  final double imageWidth;
  final BoxFit fit;
  final EdgeInsetsGeometry titlePadding;
  final EdgeInsetsGeometry descPadding;

  const OnbordingData({
    Key key,
    @required this.image,
    @required this.titleText,
    @required this.descText,
    this.imageHeight,
    this.imageWidth,
    this.fit,
    this.titlePadding,
    this.descPadding,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      alignment: Alignment.bottomCenter,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: <Widget>[

          SizedBox(
            height: 18.0,
          ),

          Padding(
              padding: titlePadding == null
                  ? EdgeInsets.symmetric(horizontal: 8.0)
                  : titlePadding,
              child: Wrap(
                children: [
                  titleText,
                ],
              )
          ),

          Image(
            fit: fit == null ? BoxFit.cover : fit,
            width: imageWidth == null ? MediaQuery.of(context).size.width : imageWidth,
            height: imageHeight == null ? MediaQuery.of(context).size.height * 0.5  : imageHeight,
            image: image,
          ),

          Padding(
              padding: descPadding == null
                  ? EdgeInsets.symmetric(horizontal: 8.0)
                  : descPadding,
              child: Wrap(
                children: [
                  descText,
                ],
              )
          ),
        ],
      ),
    );
  }
}

/// A IntroScreen Class.
class IntroScreen extends StatefulWidget {
  final List<OnbordingData> onbordingDataList;
  final MaterialPageRoute pageRoute;
  final List<Color> colors;
  final Widget nextButton;
  final Widget lastButton;
  final Widget skipButton;
  final Color selectedDotColor;
  final Color unSelectdDotColor;
  final Gradient gradient;
  IntroScreen(
      {@required this.onbordingDataList,
        this.pageRoute,
        @required this.colors,
        this.selectedDotColor,
        this.unSelectdDotColor,
        this.gradient,
        @required this.nextButton,
        @required this.lastButton,
        @required this.skipButton});

  void skipPage(BuildContext context) {
    Navigator.push(context, pageRoute);
  }

  @override
  IntroScreenState createState() {
    return new IntroScreenState();
  }
}

class IntroScreenState extends State<IntroScreen> {
  final PageController controller = new PageController();
  int currentPage = 0;
  bool lastPage = false;

  void _onPageChanged(int page) {
    setState(() {
      currentPage = page;
      if (currentPage == widget.onbordingDataList.length - 1) {
        lastPage = true;
      } else {
        lastPage = false;
      }
    });
  }

  Widget _buildPageIndicator(int page) {
    return Container(
      margin: EdgeInsets.symmetric(horizontal: 2.0),
      height: page == currentPage ? 10.0 : 6.0,
      width: page == currentPage ? 10.0 : 6.0,
      decoration: BoxDecoration(
        color: page == currentPage
            ? (widget.selectedDotColor == null
            ? Colors.blue
            : widget.selectedDotColor)
            : (widget.unSelectdDotColor == null
            ? Colors.grey
            : widget.unSelectdDotColor),
        borderRadius: BorderRadius.circular(12),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        alignment: Alignment.center,
        decoration: BoxDecoration(
          gradient: widget.gradient ==null ? LinearGradient(
              colors: [
                Colors.white,
                Colors.white,
              ]
          ) :  widget.gradient,
          color: widget.colors.length < widget.onbordingDataList.length
              ? Colors.white
              : widget.colors[currentPage],
        ),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Expanded(
              flex: 15,
              child: new PageView(
                children: widget.onbordingDataList,
                controller: controller,
                onPageChanged: _onPageChanged,
              ),
            ),
            Expanded(
              flex: 1,
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                crossAxisAlignment: CrossAxisAlignment.end,
                children: <Widget>[
                  FlatButton(
                    child: lastPage ? Text("") : widget.skipButton,
                    onPressed: () => lastPage
                        ? null
                        : widget.skipPage(
                      context,
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.symmetric(vertical: 16.0),
                    child: Container(
                      child: Row(
                          children: List.generate(
                              widget.onbordingDataList.length,
                                  (index) => _buildPageIndicator(index))),
                    ),
                  ),
                  FlatButton(
                    child: lastPage ? widget.lastButton : widget.nextButton,
                    onPressed: () => lastPage
                        ? widget.skipPage(context)
                        : controller.nextPage(
                        duration: Duration(milliseconds: 300),
                        curve: Curves.easeIn),
                  ),
                ],
              ),
            ),
            SizedBox(
              height: 8.0,
            ),
          ],
        ),
      ),
    );
  }
}




lib/flutter_walkthrough_screen.dart
// Copyright 2013 Tushar Nikam. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

library walkthrough_screen;

import 'package:flutter/material.dart';

class OnbordingData extends StatelessWidget {
  ///Contain the list of ImageProvider such as Network Asset,SVG etc Images
  final ImageProvider<Object> image;

  ///Title text
  final Widget titleText;

  ///Description of the `walkthrough`
  final Widget descText;

  /// Height of the image Default is `MediaQuery.of(context).size.height * 0.5`
  final double? imageHeight;

  ///Width of the image default is `MediaQuery.of(context).size.width`
  final double? imageWidth;

  ///Default is BoxFit.cover
  final BoxFit? fit;

  ///Default padding is `EdgeInsets.symmetric(horizontal: 8.0)`,
  final EdgeInsetsGeometry? titlePadding;

  ///Default padding is `EdgeInsets.symmetric(horizontal: 8.0)`,
  final EdgeInsetsGeometry? descPadding;

  const OnbordingData({
    Key? key,
    required this.image,
    required this.titleText,
    required this.descText,
    this.imageHeight,
    this.imageWidth,
    this.fit,
    this.titlePadding,
    this.descPadding,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      alignment: Alignment.bottomCenter,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: <Widget>[
          SizedBox(height: 18.0),
          Padding(
              padding: titlePadding == null
                  ? EdgeInsets.symmetric(horizontal: 8.0)
                  : titlePadding!,
              child: Wrap(
                children: [
                  titleText,
                ],
              )),
          Image(
            fit: fit == null ? BoxFit.cover : fit,
            width: imageWidth == null
                ? MediaQuery.of(context).size.width
                : imageWidth,
            height: imageHeight == null
                ? MediaQuery.of(context).size.height * 0.5
                : imageHeight,
            image: image,
          ),
          Padding(
              padding: descPadding == null
                  ? EdgeInsets.symmetric(horizontal: 8.0)
                  : descPadding!,
              child: Wrap(
                children: [
                  descText,
                ],
              )),
        ],
      ),
    );
  }
}

/// A IntroScreen Class.
class IntroScreen extends StatefulWidget {
  ///contain list of Onbording screen data such as
  ///OnbordingData(
  //       image: AssetImage("images/pic3.png"),
  //       fit: BoxFit.cover,
  //       imageHeight: 250,
  //       titleText: RichText(
  //         textAlign: TextAlign.center,
  //         text: TextSpan(
  //           children: [
  //             TextSpan(
  //                 text: "Pickup delivery at \nyour ",
  //                 style: TextStyle(
  //                   color: Colors.black,
  //                   fontSize: 20.0,
  //                 )),
  //             TextSpan(
  //               text: "door ! ",
  //               style: TextStyle(
  //                   color: Colors.orange,
  //                   fontSize: 22.0,
  //                   fontWeight: FontWeight.w500),
  //             ),
  //           ],
  //         ),
  //       ),
  //       descText: RichText(
  //         textAlign: TextAlign.center,
  //         text: TextSpan(
  //           children: [
  //             TextSpan(
  //               text:
  //               "Even to space with us! Together, pickup delivery at your home ",
  //               style: TextStyle(
  //                 color: Colors.grey,
  //                 fontSize: 17.0,
  //               ),
  //             ),
  //           ],
  //         ),
  //       ),
  //       descPadding: EdgeInsets.symmetric(horizontal: 22.0),
  //     ),
  final List<OnbordingData> onbordingDataList;

  ///
  final MaterialPageRoute? pageRoute;

  ///
  final List<Color> colors;

  ///
  final Widget nextButton;

  ///
  final Widget lastButton;

  ///
  final Widget skipButton;

  ///
  final Color? selectedDotColor;

  ///
  final Color? unSelectdDotColor;

  ///
  final Gradient? gradient;

  IntroScreen(
      {required this.onbordingDataList,
      this.pageRoute,
      required this.colors,
      this.selectedDotColor,
      this.unSelectdDotColor,
      this.gradient,
      required this.nextButton,
      required this.lastButton,
      required this.skipButton});

  void skipPage(BuildContext context) {
    Navigator.push(context, pageRoute!);
  }

  @override
  IntroScreenState createState() {
    return new IntroScreenState();
  }
}

class IntroScreenState extends State<IntroScreen> {
  final PageController controller = new PageController();
  int currentPage = 0;
  bool lastPage = false;

  void _onPageChanged(int page) {
    setState(() {
      currentPage = page;
      if (currentPage == widget.onbordingDataList.length - 1) {
        lastPage = true;
      } else {
        lastPage = false;
      }
    });
  }

  Widget _buildPageIndicator(int page) {
    return Container(
      margin: EdgeInsets.symmetric(horizontal: 2.0),
      height: page == currentPage ? 10.0 : 6.0,
      width: page == currentPage ? 10.0 : 6.0,
      decoration: BoxDecoration(
        color: page == currentPage
            ? (widget.selectedDotColor == null
                ? Colors.blue
                : widget.selectedDotColor)
            : (widget.unSelectdDotColor == null
                ? Colors.grey
                : widget.unSelectdDotColor),
        borderRadius: BorderRadius.circular(12),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        alignment: Alignment.center,
        decoration: BoxDecoration(
          gradient: widget.gradient == null
              ? LinearGradient(colors: [
                  Colors.white,
                  Colors.white,
                ])
              : widget.gradient,
          color: widget.colors.length < widget.onbordingDataList.length
              ? Colors.white
              : widget.colors[currentPage],
        ),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Expanded(
              flex: 15,
              child: new PageView(
                children: widget.onbordingDataList,
                controller: controller,
                onPageChanged: _onPageChanged,
              ),
            ),
            Expanded(
              flex: 1,
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                crossAxisAlignment: CrossAxisAlignment.end,
                children: <Widget>[
                  TextButton(
                    child: lastPage ? Text("") : widget.skipButton,
                    onPressed: () => lastPage
                        ? null
                        : widget.skipPage(
                            context,
                          ),
                  ),
                  Padding(
                    padding: const EdgeInsets.symmetric(vertical: 16.0),
                    child: Container(
                      child: Row(
                          children: List.generate(
                              widget.onbordingDataList.length,
                              (index) => _buildPageIndicator(index))),
                    ),
                  ),
                  TextButton(
                    child: lastPage ? widget.lastButton : widget.nextButton,
                    onPressed: () => lastPage
                        ? widget.skipPage(context)
                        : controller.nextPage(
                            duration: Duration(milliseconds: 300),
                            curve: Curves.easeIn),
                  ),
                ],
              ),
            ),
            SizedBox(
              height: 8.0,
            ),
          ],
        ),
      ),
    );
  }
}


example/example_app.dart
import 'package:flutter/material.dart';
import 'package:flutter_walkthrough_screen/flutter_walkthrough_screen.dart';
import 'next_page.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Walkthrough',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  final List<OnbordingData> list = [
    OnbordingData(
      image: AssetImage("images/pic1.png"),
      fit: BoxFit.cover,
      imageHeight: 290,
      titleText: RichText(
        textAlign: TextAlign.center,
        text: TextSpan(
          children: [
            TextSpan(
                text: "Welcome to ",
                style: TextStyle(
                  color: Colors.black,
                  fontSize: 20.0,
                )),
            TextSpan(
              text: "Foodi ",
              style: TextStyle(
                  color: Colors.orange,
                  fontSize: 22.0,
                  fontWeight: FontWeight.w500),
            ),
            TextSpan(
                text: "\nTaste us",
                style: TextStyle(
                  color: Colors.black,
                  fontSize: 20.0,
                )),
          ],
        ),
      ),
      descText: RichText(
        textAlign: TextAlign.center,
        text: TextSpan(
          children: [
            TextSpan(
              text:
                  "our app can send you everywhere even space for only \$ 2.99 per month ",
              style: TextStyle(
                color: Colors.grey,
                fontSize: 17.0,
              ),
            ),
          ],
        ),
      ),
      descPadding: EdgeInsets.symmetric(horizontal: 22.0),
    ),
    OnbordingData(
      image: AssetImage("images/pic3.png"),
      fit: BoxFit.cover,
      imageHeight: 250,
      titleText: RichText(
        textAlign: TextAlign.center,
        text: TextSpan(
          children: [
            TextSpan(
                text: "Pickup delivery at \nyour ",
                style: TextStyle(
                  color: Colors.black,
                  fontSize: 20.0,
                )),
            TextSpan(
              text: "door ! ",
              style: TextStyle(
                  color: Colors.orange,
                  fontSize: 22.0,
                  fontWeight: FontWeight.w500),
            ),
          ],
        ),
      ),
      descText: RichText(
        textAlign: TextAlign.center,
        text: TextSpan(
          children: [
            TextSpan(
              text:
                  "Even to space with us! Together, pickup delivery at your home ",
              style: TextStyle(
                color: Colors.grey,
                fontSize: 17.0,
              ),
            ),
          ],
        ),
      ),
      descPadding: EdgeInsets.symmetric(horizontal: 22.0),
    ),
    OnbordingData(
      image: AssetImage("images/pic2.png"),
      fit: BoxFit.cover,
      imageHeight: 290,
      titleText: RichText(
        textAlign: TextAlign.center,
        text: TextSpan(
          children: [
            TextSpan(
                text: "Pickup delivery at \nyour ",
                style: TextStyle(
                  color: Colors.black,
                  fontSize: 20.0,
                )),
            TextSpan(
              text: "door ! ",
              style: TextStyle(
                  color: Colors.orange,
                  fontSize: 22.0,
                  fontWeight: FontWeight.w500),
            ),
          ],
        ),
      ),
      descText: RichText(
        textAlign: TextAlign.center,
        text: TextSpan(
          children: [
            TextSpan(
              text:
                  "Even to space with us! Together, pickup delivery at your home ",
              style: TextStyle(
                color: Colors.grey,
                fontSize: 17.0,
              ),
            ),
          ],
        ),
      ),
      descPadding: EdgeInsets.symmetric(horizontal: 22.0),
    ),
  ];

  @override
  Widget build(BuildContext context) {
    return IntroScreen(
      onbordingDataList: list,
      colors: [],
      pageRoute: MaterialPageRoute(
        builder: (context) => NextPage(),
      ),
      lastButton: CircleAvatar(
        backgroundColor: Colors.indigo,
        child: Icon(
          Icons.done,
          color: Colors.white,
        ),
      ),
      nextButton: CircleAvatar(
        backgroundColor: Colors.indigo,
        child: Icon(
          Icons.arrow_forward,
          color: Colors.white,
        ),
      ),
      skipButton: Text(
        "SKIP",
        style: TextStyle(
          color: Colors.purple,
        ),
      ),
      selectedDotColor: Colors.orange,
      unSelectdDotColor: Colors.grey,
    );
  }
}
